### ISSUE TITLE
Possible static analysis or cross-compilation tools for optimizing source code before deploying to TPU?

### ISSUE THREAD

**Issue by hrbigelow** [pytorch/xla#5038]
Hi,

Is there any tool in pytorch_xla that could be run on an ordinary laptop but provide a static analysis of the python source code relevant to the target TPU?  Something like a cross-compilation?

A few things I would like to know:

1. whether there are any ops in my `nn.Module` that are producing dynamic shapes
2. if any of the ops are not lowering to xla.

And, if it's not possible to cross-compile, perhaps detecting dynamic shapes could be done anyway through a static analysis on my laptop?  I would expect that the existence of any data-dependent tensor shapes would be something that is determined entirely from the source code, and does not depend on the target device.  But maybe I'm wrong.

At the moment, my workflow is a very slow cycle of commiting local edits to my repo, pushing to gitnub, then in Colab re-installing via `!pip install` and then re-running.

Thanks in advance!

**Comment by hrbigelow** [comment]
Just to add a little context.  From [here](https://pytorch.org/xla/master/#known-performance-caveats):

>Possible sources [which cause graph re-compilations]
>Direct or indirect uses of nonzero introduce dynamic shapes; for example, masked indexing base[index] where index is a mask tensor.
>Loops with a different number of iterations between steps can result in different execution graphs, thus require recompilations.

I'm hoping a lot of these possible sources could be detected through static python source code analysis?

**Comment by JackCaoG** [comment]
for op not being lower, that's easier, you just have to do 
```
import torch_xla.debug.metrics as met

# move model and input to xla device first
loss = model(input)
loss.backward()
optimizer.step()

# do not mark_step, otherwise code will be executed.
print(met.metrics_report())
```
as long as you don't see counters like `aten::` then you are good.

For the dynamic shape part I don't have a good answer, we didn't provide any static python analysis tooling. All of our knowledge is after user actually trace the model using xla_device. The best way to check dynamic shape is to run the model for a couple steps, and check the `CompileTime` metrics. If `CompileTime` goes up as step increases, there is a dynamic shape somewhere.

**Comment by hrbigelow** [comment]
Ahh okay, thank you.  But, `CompileTime` metrics doesn't tell you which part of the python source code might be producing dynamic shapes, is that right?  You just need to search through the code and know about certain patterns?

**Comment by JackCaoG** [comment]
Take a look at https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md#more-debugging-tools, you can set 

```
XLA_IR_DEBUG=1
XLA_HLO_DEBUG=1
XLA_SAVE_TENSORS_FILE='/tmp/save.txt'
XLA_SAVE_TENSORS_FMT='text'
```
and then execute your code. This way you should see the IR dump for each graph being compiled, and if you do a diff on the graph you will see where the dynamic shape is from. Each line of IR has the top most python stack so you should be able to tell which op's input become dynamic(for example, input tensor was [2,3,4] but in step 2 it becomes [3,3,4]).

**Comment by hrbigelow** [comment]
Ahh okay, thanks.  Yes I think I've used that before a while ago.  Just was wondering if that was still the recommended way.  Thanks again.
